fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/CollectProjectAboveConsumer.java

```diff
@@ -0,0 +1,79 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.trees.plans.logical;
+
+import org.apache.doris.nereids.pattern.MatchingContext;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.rewrite.RewriteRuleFactory;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Collect Projects Above CTE Consumer.
+ */
+public class CollectProjectAboveConsumer implements RewriteRuleFactory {
+
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(RuleType.COLLECT_PROJECT_ABOVE_CONSUMER
+                        .build(logicalProject(logicalCTEConsumer()).thenApply(ctx -> {
+                            LogicalProject<LogicalCTEConsumer> project = ctx.root;
+                            List<NamedExpression> namedExpressions = project.getProjects();
+                            LogicalCTEConsumer cteConsumer = project.child();
+                            collectProject(ctx, namedExpressions, cteConsumer);
+                            return ctx.root;
+                        })),
+                RuleType.COLLECT_PROJECT_ABOVE_FILTER_CONSUMER.build(logicalProject(logicalFilter(logicalCTEConsumer()))
+                        .thenApply(ctx -> {
+                            LogicalProject<LogicalFilter<LogicalCTEConsumer>> project = ctx.root;
+                            LogicalFilter<LogicalCTEConsumer> filter = project.child();
+                            Set<Slot> filterSlots = filter.getInputSlots();
+                            List<NamedExpression> namedExpressions = new ArrayList<>(project.getProjects());
+                            for (Slot slot : filterSlots) {
+                                if (!project.getOutput().contains(slot)) {
+                                    namedExpressions.add(slot);
+                                }
+                            }
+                            collectProject(ctx, namedExpressions, filter.child());
+                            return ctx.root;
+                        }))
+        );
+    }
+
+    private static void collectProject(MatchingContext ctx,
+            List<NamedExpression> namedExpressions, LogicalCTEConsumer cteConsumer) {
+        for (Expression expr : namedExpressions) {
+            expr.foreach(node -> {
+                if (!(node instanceof Slot)) {
+                    return;
+                }
+                Slot slot = cteConsumer.findProducerSlot((Slot) node);
+                ctx.cascadesContext.putCTEIdToProject(cteConsumer.getCteId(), slot);
+                ctx.cascadesContext.markConsumerUnderProject(cteConsumer);
```

Comment:
put some in ctx but not use, just remove this line

Reply:
ditto